pkg/controller/job/job_controller.go

```diff
@@ -669,140 +670,171 @@ func newCondition(conditionType batch.JobConditionType, reason, message string)
-					// Decrement the expected number of creates because the informer won't observe this pod
-					jm.expectations.CreationObserved(jobKey)
+
+		} else if active < parallelism {
+			wantActive := int32(0)
+			if job.Spec.Completions == nil {
+				// Job does not specify a number of completions.  Therefore, number active
+				// should be equal to parallelism, unless the job has seen at least
+				// once success, in which leave whatever is running, running.
+				if succeeded > 0 {
+					wantActive = active
+				} else {
+					wantActive = parallelism
 				}
-				// The skipped pods will be retried later. The next controller resync will
-				// retry the slow start process.
-				break
+			} else {
+				// Job specifies a specific number of completions.  Therefore, number
+				// active should not ever exceed number of remaining completions.
+				wantActive = *job.Spec.Completions - succeeded
+				if wantActive > parallelism {
+					wantActive = parallelism
+				}
+			}
+			diff := wantActive - active
+			if diff < 0 {
+				utilruntime.HandleError(fmt.Errorf("More active than wanted: job %q, want %d, have %d", jobKey, wantActive, active))
+				diff = 0
+			}
+			jm.expectations.ExpectCreations(jobKey, int(diff))
+			errCh = make(chan error, diff)
+			glog.V(4).Infof("Too few pods running job %q, need %d, creating %d", jobKey, wantActive, diff)
+
+			allocation := sync.Mutex{}
+			availableCompletionsIndexes := make([]int32, 0)
+			if completions > 0 {
+				// find available completions indexes in now
+				availableCompletionsIndexes = getAvailableCompletionsIndexes(activePods, succeededPods, completions)
+			}
+
+			active += diff
+			wait := sync.WaitGroup{}
+
+			// Batch the pod creates. Batch sizes start at SlowStartInitialBatchSize
+			// and double with each successful iteration in a kind of "slow start".
+			// This handles attempts to start large numbers of pods that would
+			// likely all fail with the same error. For example a project with a
+			// low quota that attempts to create a large number of pods will be
+			// prevented from spamming the API service with the pod create requests
+			// after one of its pods fails.  Conveniently, this also prevents the
+			// event spam that those failures would generate.
+			for batchSize := int32(integer.IntMin(int(diff), controller.SlowStartInitialBatchSize)); diff > 0; batchSize = integer.Int32Min(2*batchSize, diff) {
+				errorCount := len(errCh)
+				wait.Add(int(batchSize))
+				for i := int32(0); i < batchSize; i++ {
+					go func() {
+						defer wait.Done()
+						completionsIndex := int32(0)
+						if completions > 0 {
+							// allocation completions index
+							allocation.Lock()
+							if len(availableCompletionsIndexes) > 0 {
+								completionsIndex = availableCompletionsIndexes[0]
+								availableCompletionsIndexes = availableCompletionsIndexes[1:]
+							}
+							allocation.Unlock()
+							if completionsIndex == 0 {
+								// TODO log warning
```

Comment:
can the log warning be added as part of this PR?

Reply:
ok